File: app/src/main/java/br/com/jonathanzanella/myexpenses/bill/BillPresenter.java
package br.com.jonathanzanella.myexpenses.bill;

import android.app.DatePickerDialog;
import android.content.Context;
import android.os.AsyncTask;
import android.support.annotation.UiThread;
import android.support.annotation.WorkerThread;
import android.widget.DatePicker;

import org.joda.time.DateTime;

import br.com.jonathanzanella.myexpenses.R;
import br.com.jonathanzanella.myexpenses.exceptions.InvalidMethodCallException;
import br.com.jonathanzanella.myexpenses.helpers.CountingIdlingResource;
import br.com.jonathanzanella.myexpenses.validations.OperationResult;
import br.com.jonathanzanella.myexpenses.validations.ValidationError;

/**
 * Created by jzanella on 8/27/16.
 */

class BillPresenter {
	private BillContract.View view;
	private BillContract.EditView editView;
	private final BillRepository repository;
	private Bill bill;
	private DateTime initDate;
	private DateTime endDate;
	private final CountingIdlingResource idlingResource;

	BillPresenter(BillRepository repository, CountingIdlingResource idlingResource) {
		this.repository = repository;
		this.idlingResource = idlingResource;
	}

	void attachView(BillContract.View view) {
		this.view = view;
	}

	void attachView(BillContract.EditView view) {
		this.view = view;
		this.editView = view;
	}

	void detachView() {
		this.view = null;
		this.editView = null;
	}

	@UiThread
	void onViewUpdated(boolean invalidateCache) {
		if (bill != null) {
			if(invalidateCache) {
				new AsyncTask<Void, Void, Void>() {

					@Override
					protected Void doInBackground(Void... voids) {
						loadBill(bill.getUuid());
						return null;
					}

					@Override
					protected void onPostExecute(Void aVoid) {
						super.onPostExecute(aVoid);
						updateView();
					}
				}.execute();
			}
		} else {
			if(editView != null)
				editView.setTitle(R.string.new_bill_title);
		}
	}

	private void updateView() {
		if(editView != null) {
			editView.setTitle(R.string.edit_bill_title);
		} else {
			String title = view.getContext().getString(R.string.bill);
			view.setTitle(title.concat(" ").concat(bill.getName()));
		}
		view.showBill(bill);
		initDate = bill.getInitDate();
		if(editView != null)
			editView.onInitDateChanged(initDate);
		endDate = bill.getEndDate();
		if(editView != null)
			editView.onEndDateChanged(endDate);
	}

	void onInitDate(Context ctx) {
		checkEditViewSet();
		DateTime time = initDate;
		if(time == null)
			time = DateTime.now();
		new DatePickerDialog(ctx, new DatePickerDialog.OnDateSetListener() {
			@Override
			public void onDateSet(DatePicker view, int year, int monthOfYear, int dayOfMonth) {
				if(initDate == null)
					initDate = DateTime.now();
				initDate = initDate.withYear(year).withMonthOfYear(monthOfYear + 1).withDayOfMonth(dayOfMonth);
				editView.onInitDateChanged(initDate);
			}
		}, time.getYear(), time.getMonthOfYear() - 1, time.getDayOfMonth()).show();
	}

	void onEndDate(Context ctx) {
		checkEditViewSet();
		DateTime time = initDate;
		if(time == null)
			time = DateTime.now();
		new DatePickerDialog(ctx, new DatePickerDialog.OnDateSetListener() {
			@Override
			public void onDateSet(DatePicker view, int year, int monthOfYear, int dayOfMonth) {
				if(endDate == null)
					endDate = DateTime.now();
				endDate = endDate.withYear(year).withMonthOfYear(monthOfYear + 1).withDayOfMonth(dayOfMonth);
				editView.onEndDateChanged(endDate);
			}
		}, time.getYear(), time.getMonthOfYear() - 1, time.getDayOfMonth()).show();
	}

	@WorkerThread
	void loadBill(final String uuid) {
		bill = repository.find(uuid);
	}

	private void checkEditViewSet() {
		if(editView == null)
			throw new InvalidMethodCallException("save", getClass().toString(), "View should be a Edit View");
	}

	void save() {
		checkEditViewSet();
		if(bill == null)
			bill = new Bill();
		bill = editView.fillBill(bill);
		bill.setInitDate(initDate);
		bill.setEndDate(endDate);
		OperationResult result = repository.save(bill);

		if(result.isValid()) {
			editView.finishView();
		} else {
			for (ValidationError validationError : result.getErrors())
				editView.showError(validationError);
		}
	}

	String getUuid() {
		return bill != null ? bill.getUuid() : null;
	}
}

File: app/src/main/java/br/com/jonathanzanella/myexpenses/bill/BillPresenter.kt
package br.com.jonathanzanella.myexpenses.bill

import android.app.DatePickerDialog
import android.content.Context
import android.os.AsyncTask
import android.support.annotation.UiThread
import android.support.annotation.WorkerThread
import br.com.jonathanzanella.myexpenses.R
import br.com.jonathanzanella.myexpenses.exceptions.InvalidMethodCallException
import org.joda.time.DateTime

class BillPresenter(private val repository: BillRepository) {
    private var view: BillContract.View? = null
    private var editView: BillContract.EditView? = null
    private var bill: Bill? = null
    private var initDate: DateTime? = null
    private var endDate: DateTime? = null

    fun attachView(view: BillContract.View) {
        this.view = view
    }

    fun attachView(view: BillContract.EditView) {
        this.view = view
        this.editView = view
    }

    fun detachView() {
        this.view = null
        this.editView = null
    }

    @UiThread
    fun onViewUpdated(invalidateCache: Boolean) {
        if (bill != null) {
            if (invalidateCache) {
                object : AsyncTask<Void, Void, Void>() {

                    override fun doInBackground(vararg voids: Void): Void? {
                        loadBill(bill!!.uuid!!)
                        return null
                    }

                    override fun onPostExecute(aVoid: Void?) {
                        super.onPostExecute(aVoid)
                        updateView()
                    }
                }.execute()
            } else {
                updateView()
            }
        } else {
            updateView()
        }
    }

    fun updateView() {
        val v = editView
        val b = bill
        if (b != null) {
            if (v != null) {
                v.setTitle(R.string.edit_bill_title)
            } else {
                view!!.let {
                    val title = it.context.getString(R.string.bill)
                    it.setTitle(title + " " + b.name)
                }
            }
            view!!.showBill(b)
            initDate = b.initDate
            v?.onInitDateChanged(initDate!!)
            endDate = b.endDate
            v?.onEndDateChanged(endDate!!)
        } else {
            v?.setTitle(R.string.new_bill_title)
        }
    }

    fun onInitDate(ctx: Context) {
        checkEditViewSet()
        var time = initDate
        if (time == null)
            time = DateTime.now()
        DatePickerDialog(ctx, DatePickerDialog.OnDateSetListener { view, year, monthOfYear, dayOfMonth ->
            if (initDate == null)
                initDate = DateTime.now()
            initDate = initDate!!.withYear(year).withMonthOfYear(monthOfYear + 1).withDayOfMonth(dayOfMonth)
            editView!!.onInitDateChanged(initDate!!)
        }, time!!.year, time.monthOfYear - 1, time.dayOfMonth).show()
    }

    fun onEndDate(ctx: Context) {
        checkEditViewSet()
        var time = initDate
        if (time == null)
            time = DateTime.now()
        DatePickerDialog(ctx, DatePickerDialog.OnDateSetListener { view, year, monthOfYear, dayOfMonth ->
            if (endDate == null)
                endDate = DateTime.now()
            endDate = endDate!!.withYear(year).withMonthOfYear(monthOfYear + 1).withDayOfMonth(dayOfMonth)
            editView!!.onEndDateChanged(endDate!!)
        }, time!!.year, time.monthOfYear - 1, time.dayOfMonth).show()
    }

    @WorkerThread
    fun loadBill(uuid: String) {
        bill = repository.find(uuid)
    }

    private fun checkEditViewSet() {
        if (editView == null)
            throw InvalidMethodCallException("save", javaClass.toString(), "View should be a Edit View")
    }

    fun save() {
        checkEditViewSet()

        val v = editView!!
        val b = when(bill) {
            null -> v.fillBill(Bill())
            else -> v.fillBill(bill!!)
        }
        b.initDate = initDate
        b.endDate = endDate
        val result = repository.save(b)

        if (result.isValid) {
            v.finishView()
        } else {
            for (validationError in result.errors)
                v.showError(validationError)
        }
    }

    val uuid: String?
        get() = bill?.uuid
}
